[tests] Fix intermittent rpc_net.py failure.

jnewbery opened the pull request:
~rpc_net.py would intermittently fail on Travis, probably~
~due to assuming that two consecutive RPC calls were atomic.~
~Fix this by trying the test up to three times and only failing~
~if all three attempts fail.~

EDIT: PR updated to:

rpc_net.py would intermittently fail on Travis, probably
due to assuming that two consecutive RPC calls were atomic.
Fix this by only testing that amounts are bounded above and
below rather than equal.

fixes #11778

jnewbery committed:
[tests] Fix intermittent rpc_net.py failure.

rpc_net.py would intermittently fail on Travis, probably
due to assuming that two consecutive RPC calls were atomic.
Fix this by only testing that amounts are bounded above and
below rather than equal.

conscott commented on test/functional/rpc_net.py:
```diff
@@ -33,15 +33,25 @@ def _test_connection_count(self):
+        # Check that getnettotals totalbytesrecv and totalbytessent
+        # are consistent with getpeerinfo. Since the RPC calls are
+        # not atomic, this may fail. Try up to 3 times.
+        for i in range(3):
```
Does it make sense to add a `time.sleep()` before `getpeerinfo()` so that the two calls might have a better chance of being in sync?

I am also curious if you have been able to reproduce locally; I have not. I assume this changes the likelihood of failure significantly lower, but I am curious if in the case where the first try fails, might the second also be likely to fail for whatever reason.

jnewbery commented:
ok. I've thought about this some more, and changed the approach. Rather than trying three times and hoping that there'll be equality at some point, I'm now calling `getnettotals` before and after `getpeerinfo` and asserting that the info returned by `getpeerinfo` is bounded above and below by the totals in `getnettotals`. I've also changed some equality tests below to inequality tests, for the same reason - that there may be unexpected network traffic between the two calls.

We can't guarantee that there won't be network traffic between two separate RPC calls, so we shouldn't be testing for that.

MeshCollider commented:
I like the second approach much better than the 3 attempts, utACK https://github.com/bitcoin/bitcoin/pull/12804/commits/5a67c0524e5dc98d0e387f189545bc99863916d4

MarcoFalke commented:
utACK 5a67c0524e5dc98d0e387f189545bc99863916d4

promag commented:
PR description invalid?

utACK 5a67c05.

promag commented:
Could still copy commit message to PR description so it's correct for future reference.

jnewbery commented:
> Could still copy commit message to PR description so it's correct for future reference.

Sorry - missed this entirely. PR description now fixed.
